straxen/plugins/acqmon_processing.py

```diff
@@ -90,44 +94,125 @@ def setup(self):
-def channel_select(rr, ch_stop, ch_start):
-    """Return data in between stop and start channels of the acquisition monitor (AM)"""
-    return rr[((rr['channel'] == ch_stop) | (rr['channel'] == ch_start))] 
+def channel_select_(rr, ch):
+    """Return data from start/stop veto channel in the acquisition monitor (AM)"""
+    return rr[rr['channel'] == ch] 
+
+
+
+
+@export
+@strax.takes_config(
+    strax.Option('veto_proximity_window', default=int(5e8), type=int, track=True,
+                 help='Maximum separation between veto stop and start pulses [ns]'))
+
+class VetoProximity(strax.OverlapWindowPlugin):
+    """ 
+    Find the closest next/previous veto start and end to each event center.
+    
+    previous_x: Time in ns between the time center of an event and the previous x
+    next_x: Time in ns between the time center of an event and the next x
+    This also considers any x inside the event. x could be either:
+        - busy_x: busy on/off signal
+        - he_x:   high energy channels busy on/off signal
+        - hev_x:  high energy veto on/off signal
+    """
+        
+    __version__ = '0.0.2'
+    depends_on = ('veto_intervals','events') 
+    provides  = ('veto_proximity')
+    data_kind = ('veto_proximity')
+    
+    veto_names = ['busy', 'he', 'hev']
+    
+    save_when = strax.SaveWhen.NEVER
+    
+    def infer_dtype(self):
+        dtype = []
+        for n in self.veto_names:
+            dtype.append((('Time to previous %s veto start [ns]' %n, 'previous_%s_on' %n), np.int64))
+            dtype.append((('Time to previous %s veto end [ns]' %n, 'previous_%s_off' %n), np.int64))
+            dtype.append((('Time to next %s veto start [ns]' %n, 'next_%s_on' %n), np.int64))
+            dtype.append((('Time to next %s veto end [ns]' %n, 'next_%s_off' %n), np.int64))
+        
+        dtype += strax.time_fields
+
+        return dtype
+    
+    def setup(self):
+        self.states = ['on', 'off']
+    
+    def get_window_size(self):
+        return (self.config['veto_proximity_window'] * 100)
+        
+    def compute(self, events, veto_intervals):
+        res = defaultdict(list)
+        
+        t_event_centers = (events['time'] + events['endtime'])//2
+
+        for n, name in enumerate(self.veto_names):
+            veto_selection = veto_intervals[veto_intervals['veto_type'] == ('%s_veto' %name)]
+            
+            # For each state find the next and previous veto
+            for state in self.states:
+                prev = 'previous_%s_%s' %(name, state)
+                nxt = 'next_%s_%s' %(name, state)
+                
+                if state == 'on':
+                    veto_start_time_selection = veto_selection['time']
+                else:
+                    veto_start_time_selection = veto_selection['endtime']
+                    
+                for t, time in enumerate(t_event_centers):
+                    inx = np.searchsorted(veto_start_time_selection, time, side = 'right')
```

Comment:
I think that at this point, you can also add:
`inx = np.searchsorted(veto_start_time_selection[prev_inx:], time, side = 'right')`
whereafter you set `prev_inx` to `inx` since the events are sorted in time. That way you don't have to each the first x veto intervals if you know that previous event wasn't in those first x either.

I see that most likely there is little events so it will be fast anyway we are doing a third level for loop here. Faster still would be if we do the for loop of the names and states inside the for loop of time such that you only have to traverse the time loop (longest) once. But I guess this also has it advantage that it looks rather clear what you are doing.

Reply:
Cheers, for the suggestions, if there would be some speed issues for the bigger runs I'd update this accordingly. :)